include/pybind11/eigen.h

```diff
@@ -641,6 +643,343 @@ struct type_caster<Type, enable_if_t<is_eigen_other<Type>::value>> {
+template <typename Type>
+struct type_caster<Eigen::TensorMap<Type>, typename eigen_helper<Type>::ValidType> {
+    using H = eigen_helper<Type>;
+
+    bool load(handle src, bool /*convert*/) {
+        // Note that we have a lot more checks here as we want to make sure to avoid copies
+        auto a = reinterpret_borrow<array>(src);
+        if ((a.flags() & eigen_to_numpy<H::Options>::flag) == 0) {
+            return false;
+        }
+
+        if (!a.dtype().is(dtype::of<typename H::Element>())) {
+            return false;
+        }
+
+        if (a.ndim() != H::N) {
+            return false;
+        }
+
+        std::array<EigenIndex, H::N> shape;
+        std::copy(a.shape(), a.shape() + H::N, shape.begin());
+
+        if (!H::is_correct_shape(shape)) {
+            return false;
+        }
+
+        value.reset(new Eigen::TensorMap<Type>(
+            reinterpret_cast<typename H::Element *>(a.mutable_data()), shape));
+
+        return true;
+    }
+
+    static handle cast(Eigen::TensorMap<Type> &&src, return_value_policy policy, handle parent) {
+        return cast_impl(&src, policy, parent);
+    }
+
+    static handle
+    cast(const Eigen::TensorMap<Type> &&src, return_value_policy policy, handle parent) {
+        return cast_impl(&src, policy, parent);
+    }
+
+    static handle cast(Eigen::TensorMap<Type> &src, return_value_policy policy, handle parent) {
+        if (policy == return_value_policy::automatic
+            || policy == return_value_policy::automatic_reference) {
+            policy = return_value_policy::copy;
+        }
+        return cast_impl(&src, policy, parent);
+    }
+
+    static handle
+    cast(const Eigen::TensorMap<Type> &src, return_value_policy policy, handle parent) {
+        if (policy == return_value_policy::automatic
+            || policy == return_value_policy::automatic_reference) {
+            policy = return_value_policy::copy;
+        }
+        return cast(&src, policy, parent);
+    }
+
+    static handle cast(Eigen::TensorMap<Type> *src, return_value_policy policy, handle parent) {
+        if (policy == return_value_policy::automatic) {
+            policy = return_value_policy::take_ownership;
+        } else if (policy == return_value_policy::automatic_reference) {
+            policy = return_value_policy::reference;
+        }
+        return cast_impl(src, policy, parent);
+    }
+
+    static handle
+    cast(const Eigen::TensorMap<Type> *src, return_value_policy policy, handle parent) {
+        if (policy == return_value_policy::automatic) {
+            policy = return_value_policy::take_ownership;
+        } else if (policy == return_value_policy::automatic_reference) {
+            policy = return_value_policy::reference;
+        }
+        return cast_impl(src, policy, parent);
+    }
+
+    template <typename C>
+    static handle cast_impl(C *src, return_value_policy policy, handle parent) {
+        bool dec_parent = false;
+        bool writeable = false;
```

Comment:
Writeable can and should be definitely be a constexpr boolean in this case (only dependent on std::is_const).